bpf/dns_tracker.h

```diff
@@ -0,0 +1,141 @@
+            new_flow.start_mono_time_ts = current_time;
+            new_flow.end_mono_time_ts = current_time;
+            new_flow.packets = 1;
+            new_flow.bytes = len;
+            new_flow.flags = flags;
+            new_flow.dns_record.id = bpf_ntohs(dns->id);
+            new_flow.dns_record.flags = bpf_ntohs(dns->flags);
+            new_flow.dns_record.req_mono_time_ts = current_time;
+        } else {
+            new_flow.start_mono_time_ts = current_time;
+            new_flow.end_mono_time_ts = current_time;
+            new_flow.packets = 1;
+            new_flow.bytes = len;
+            new_flow.flags = flags;
+            new_flow.dns_record.id = bpf_ntohs(dns->id);
+            new_flow.dns_record.flags = bpf_ntohs(dns->flags);
+            new_flow.dns_record.rsp_mono_time_ts = current_time;
+        }
+        bpf_map_update_elem(&aggregated_flows, id, &new_flow, BPF_ANY);
+    }
+}
+
+static inline int trace_dns(struct sk_buff *skb) {
+    struct sock *sk = skb->sk;
+    flow_id id;
+    u8 protocol = 0;
+    u16 family = 0,flags = 0, len = 0;
+
+    __builtin_memset(&id, 0, sizeof(id));
+
+    // pull in details from the packet headers and the sock struct
+    bpf_probe_read(&family, sizeof(u16), &sk->__sk_common.skc_family);
+
+    id.if_index = skb->skb_iif;
+
+    // read L2 info
+    set_key_with_l2_info(skb, &id);
+
+    // read L3 info
+    set_key_with_l3_info(skb, family, &id, &protocol);
+
+    switch (protocol) {
+    case IPPROTO_UDP:
+        len = set_key_with_udp_info(skb, &id, IPPROTO_UDP);
+        break;
+    case IPPROTO_TCP:
+        len = set_key_with_tcp_info(skb, &id, IPPROTO_TCP, &flags);
+        break;
+    default:
+        return -1;
+    }
+
+    // check for DNS packets
+    if (id.dst_port == DNS_PORT || id.src_port == DNS_PORT) {
+        struct dns_header dns;
+        bpf_probe_read(&dns, sizeof(dns), (struct dns_header *)(skb->head + skb->transport_header + len));
+
+        struct dns_request_key req_key = {
+            .id = dns.id
+        };
+        struct dns_request_val dns_val;
+
+        __builtin_memset(&dns_val, 0, sizeof(dns_val));
+
+        if ((dns.flags & DNS_QR_FLAG) == 0) { /* dns query */
+            if (bpf_map_lookup_elem(&dns_requests, &req_key) == NULL) {
+                __builtin_memcpy(dns_val.srcip, id.src_ip, IP_MAX_LEN);
+                bpf_map_update_elem(&dns_requests, &req_key, &dns_val, BPF_ANY);
+                find_or_create_dns_flow(&id , &dns, skb->len, INGRESS, flags);
+            }
+        } else { /* dns response */
+            struct dns_request_val *value = bpf_map_lookup_elem(&dns_requests, &req_key);
+            if (value != 0 && (dns_addr_compare(value->srcip, id.dst_ip, IP_MAX_LEN) == 0)) {
+                find_or_create_dns_flow(&id , &dns, skb->len, EGRESS, flags);
+            }
+        } // end of dns response
+    } // end of dns port check
+
+    return 0;
+}
+
+SEC("tracepoint/net/net_dev_queue")
```

Comment:
This means you'll most likely see each packet multiple times; I assume that's fine for the use case?

Reply:
we look for the 1st DNS query then wait for matching response then delete that record, so if we see multiple DNS reqq we will just take the 1st one and ignore the rest

Reply:
Right, there's no guarantee you won't get collisions, though? ATAICT you're not using the src/dst IP and port as part of the DNS key, so you just have the 16-bit DNS ID as a globally unique identifier? 16 bits is not a lot, so you'll definitely get collisions this way...

Also, is there a mechanism for cleaning out the map of entries that never get a response somewhere?

Reply:
its LRU map so I assume the stale entry will get magically cleared out by kernel no ?